sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/PaginationCarbonReader.java

```diff
@@ -0,0 +1,303 @@
+
+  /**
+   * This interface is for python to call java.
+   * Because python cannot understand java Long object. so send string object.
+   *
+   * Get total rows in the folder or a list of CarbonData files.
+   * It is based on the snapshot of files taken while building the reader.
+   *
+   *
+   * @return total rows from all the files in the reader.
+   */
+  public String getTotalRowsAsString() {
+    if (isClosed) {
+      throw new RuntimeException("Pagination Reader is closed. please build again");
+    }
+    return (rowCountInSplits.get(rowCountInSplits.size() - 1)).toString();
+  }
+
+  private static int findBlockletIndex(List<Long> summationArray, Long key) {
+    // summation array is in sorted order, so can use the binary search.
+    int index = Collections.binarySearch(summationArray, key);
+    if (index < 0) {
+      // when key not found, binary search returns negative index [-1 to -N].
+      // which is the possible place where key can be inserted.
+      // with one shifted position. As 0 is also a valid index.
+      // offset the one index shifted and get absolute value of it.
+      index = Math.abs(index + 1);
+    }
+    return index;
+  }
+
+  private Range getBlockletIndexRange(long fromRowNumber, long toRowNumber) {
+    // find the matching blocklet index range by checking with from and to.
+    int upperBound = findBlockletIndex(rowCountInSplits, toRowNumber);
+    // lower bound cannot be more than upper bound, so work on sub list.
+    int lowerBound = findBlockletIndex(rowCountInSplits.subList(0, upperBound), fromRowNumber);
+    return new Range(lowerBound, upperBound);
+  }
+
+  private Object[] getRows(long fromRowNumber, long toRowNumber)
+      throws IOException, InterruptedException {
+    int rowCount = 0;
+    Object[] rows = new Object[(int)(toRowNumber - fromRowNumber + 1)];
+    // get the matching split index (blocklets) range for the input range.
+    Range blockletIndexRange = getBlockletIndexRange(fromRowNumber, toRowNumber);
+    for (int i = blockletIndexRange.getFrom(); i <= blockletIndexRange.getTo(); i++) {
+      String blockletUniqueId = String.valueOf(i);
+      BlockletRows blockletRows;
+      if (cache.get(blockletUniqueId) != null) {
+        blockletRows = (BlockletRows)cache.get(blockletUniqueId);
+      } else {
+        BlockletDetailInfo detailInfo =
+            ((CarbonInputSplit) allBlockletSplits.get(i)).getDetailInfo();
+        int rowCountInBlocklet = detailInfo.getRowCount();
+        Object[] rowsInBlocklet = new Object[rowCountInBlocklet];
+        // read the rows from the blocklet
+        // TODO: read blocklets in multi-thread if there is a performance requirement.
+        readerBuilder.setInputSplit(allBlockletSplits.get(i));
+        CarbonReader<Object> carbonReader = readerBuilder.build();
+        int count = 0;
+        while (carbonReader.hasNext()) {
+          rowsInBlocklet[count++] = carbonReader.readNextRow();
+        }
+        carbonReader.close();
```

Comment:
hasNext() seem to throw IOException, InterruptedException exceptions. Enclose this while loop in try/catch/finally block is better ?

Reply:
If we need to do something when the exception occurred it make sense to enclose it. Now we don't have anything to do.